src/page/page_cache.rs

```diff
@@ -1,50 +1,65 @@
-use super::*;
+use std::io::{Read, Seek, Write};
 
+use crossbeam::sync::AtomicOption;
 use rayon::prelude::*;
+use zstd::block::{compress, decompress};
+
+use super::*;
 
 /// A lock-free pagecache.
-pub struct PageCache<L: Log, M>
-    where M: Materializer + Sized,
-          L: Log + Sized
-{
-    t: M,
-    inner: Radix<Stack<CacheEntry<M>>>,
-    max_id: AtomicUsize,
+pub struct PageCache<PM, L, P, R> {
+    t: PM,
+    inner: Radix<Stack<CacheEntry<P>>>,
+    max_pid: AtomicUsize,
     free: Stack<PageID>,
-    log: Box<L>,
+    log: L,
     lru: lru::Lru,
+    updates: AtomicUsize,
+    last_snapshot: AtomicOption<Snapshot<R>>,
```

Comment:
we store the last snapshot so that when we generate a new one, we can create a diff by scanning the log between the last snapshot point and the tip.